lib/datura/elasticsearch/index.rb

```diff
@@ -0,0 +1,260 @@
+        # great, the field is valid, now check if it is a parent
+        Array(value).each do |nested|
+          if nested.class == Hash
+            if nested.keys.all? { |k| valid_field?(k, field) }
+              next
+            else
+              # if one of the nested hashes fails, it is invalid
+              puts "Nested field '#{field}' is invalid"
+              return false
+            end
+          end
+        end
+        # all nested fields passed, so it is valid
+        true
+      else
+        puts "Field '#{field}' is invalid"
+        false
+      end
+    end
+  end
+
+  # if a field, including those inside nested fields,
+  # matches a top level field mapping or a dynamic field,
+  # they are good to go
+  # further, if this is a nested field, they may check
+  # to see if the specific nesting mapping validates them
+  def valid_field?(field, parent=nil)
+    @schema_mapping["fields"].include?(field) ||
+      field.match(@schema_mapping["dynamic"]) ||
+      valid_nested_field?(field, parent)
+  end
+
+  def valid_nested_field?(field, parent)
+    parent_mapping = @schema_mapping["nested"][parent]
+    parent_mapping.include?(field) if parent_mapping
+  end
+
+  def self.clear
+    # run the parameters through the option parser
+    params = Datura::Parser.clear_index
+    options = Datura::Options.new(params).all
+    if options["collection"] == "all"
+      self.clear_all(options)
+    else
+      self.clear_index(options)
+    end
+  end
+
+  private
+
+  def self.build_clear_data(options)
+    if options["regex"]
+      field = options["field"] || "identifier"
+      {
+        "query" => {
+          "bool" => {
+            "must" => [
+              { "regexp" => { field => options["regex"] } },
+              { "term" => { "collection" => options["collection"] } }
+            ]
+          }
+        }
+      }
+    else
+      {
+        "query" => { "term" => { "collection" => options["collection"] } }
+      }
+    end
+  end
+
+  def self.clear_all(options)
+    puts "Please verify that you want to clear EVERY ENTRY from the ENTIRE INDEX\n\n"
+    puts "== FIELD / REGEX FILTERS NOT AVAILABLE FOR THIS OPTION, YOU'LL WIPE EVERYTHING ==\n\n"
+    puts "Running this on something other than your computer's localhost? DON'T."
+    puts "Type: 'Yes I'm sure'"
+    confirm = STDIN.gets.chomp
+    if confirm == "Yes I'm sure"
+      url = File.join(options["es_path"], options["es_index"], "_delete_by_query?pretty=true")
+      auth_header = Datura::Helpers.construct_auth_header(options)
+      json = { "query" => { "match_all" => {} } }
+      RestClient.post(url, json.to_json, auth_header.merge({ content_type: :json })) { |res, req, result|
+        if result.code == "200"
+          puts res
+        else
+          raise "#{result.code} error when clearing entire index: #{res}"
+        end
+      }
+    else
+      puts "You typed '#{confirm}'. This is incorrect, exiting program"
+      exit
+    end
+  end
```

Comment:
It looks like this code is duplicated between this file `index.rb` and `data.rb` for the Elasticsearch module. Please check which methods are duplicate and if they are both actually in use via the commands in `bin/` or not

Reply:
`index.rb` is where the method is called. I don't think `data.rb` is called at all.

Reply:
deleted